sound/soc/sdca/sdca_interrupts.c

```diff
@@ -0,0 +1,193 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * SDCA IRQ handler library
+ *
+ * Copyright (C) 2025 Cirrus Logic, Inc. and
+ *                         Cirrus Logic International Semiconductor Ltd.
+ *
+ * The MIPI SDCA specification is available for public downloads at
+ * https://www.mipi.org/mipi-sdca-v1-0-download
+ */
+#include <linux/device.h>
+#include <linux/soundwire/sdw.h>
+#include <linux/soundwire/sdw_registers.h>
+#include <sound/sdca.h>
+#include <linux/regmap.h>
+#include <linux/interrupt.h>
+#include <linux/workqueue.h>
+#include <sound/sdca_interrupts.h>
+#include <sound/sdca_function.h>
+
+static irqreturn_t sdca_base_irq_handler(int irq, void *data)
+{
+	struct sdca_interrupt *priv = data;
+
+	return IRQ_HANDLED;
+}
+
+#define IRQ_SDCA_OFFSET(reg_number) ((SDW_SCP_SDCA_INT##reg_number) - SDW_SCP_SDCA_INT1)
+#define IRQ_SDCA(number) REGMAP_IRQ_REG(number, \
+						    IRQ_SDCA_OFFSET((number / 8) + 1), \
+						    SDW_SCP_SDCA_INTMASK_SDCA_##number)
+
+static const struct regmap_irq regmap_irqs[] = {
+	IRQ_SDCA(0),
+	IRQ_SDCA(1),
+	IRQ_SDCA(2),
+	IRQ_SDCA(3),
+	IRQ_SDCA(4),
+	IRQ_SDCA(5),
+	IRQ_SDCA(6),
+	IRQ_SDCA(7),
+	IRQ_SDCA(8),
+	IRQ_SDCA(9),
+	IRQ_SDCA(10),
+	IRQ_SDCA(11),
+	IRQ_SDCA(12),
+	IRQ_SDCA(13),
+	IRQ_SDCA(14),
+	IRQ_SDCA(15),
+	IRQ_SDCA(16),
+	IRQ_SDCA(17),
+	IRQ_SDCA(18),
+	IRQ_SDCA(19),
+	IRQ_SDCA(20),
+	IRQ_SDCA(21),
+	IRQ_SDCA(22),
+	IRQ_SDCA(23),
+	IRQ_SDCA(24),
+	IRQ_SDCA(25),
+	IRQ_SDCA(26),
+	IRQ_SDCA(27),
+	IRQ_SDCA(28),
+	IRQ_SDCA(29),
+	IRQ_SDCA(30),
+};
+
+static const struct regmap_irq_chip sdca_irq_chip = {
+	.name = "sdca_irq",
+
+	.status_base = SDW_SCP_SDCA_INT1,
+	.unmask_base = SDW_SCP_SDCA_INTMASK1,
+	.ack_base = SDW_SCP_SDCA_INT1,
+	.num_regs = 4,
+
+	.irqs = regmap_irqs,
+	.num_irqs = SDCA_MAX_INTERRUPTS,
+
+	.runtime_pm = true,
+};
+
+int sdca_request_irq(struct sdca_interrupt_info *interrupt_info, struct sdca_interrupt *irq)
+{
+	char *name = irq->name;
+	irq_handler_t thread_fn = irq->callback;
+	int ret;
+
+	ret = irq_create_mapping(interrupt_info->irq_dom, irq->num);
+
+	if (ret < 0) {
+		dev_err(interrupt_info->dev, "Failed to map IRQ %s\n", name);
+		return 1;
+	}
+
+	dev_dbg(interrupt_info->dev, "Request IRQ %d for %s\n", ret, name);
+
+	ret = devm_request_threaded_irq(interrupt_info->dev, ret, NULL, thread_fn,
+					IRQF_ONESHOT, name, irq);
+	if (ret) {
+		dev_err(interrupt_info->dev, "Failed to request IRQ %s\n", name);
+		return 1;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(sdca_request_irq);
+
+static int sdca_process_controls_with_interrupts(struct sdca_interrupt_info *interrupt_info,
+						 struct sdca_function_data *functions,
+						 int num_functions)
```

Comment:
Continued... then this can become something that gets called from each function driver, so it no longer needs to manually loop over all the functions and is a little more in sync with the other APIs. I would also be inclined to split it into 2 functions one that returns an irq count for the function and that populates the structs. That is in keeping with all the asoc SDCA apis keeps everything consistent.